src/pybind/mgr/dashboard/controllers/grafana.py

```diff
@@ -16,21 +17,26 @@
 @ApiController('/grafana', Scope.GRAFANA)
 @ControllerDoc("Grafana Management API", "Grafana")
 class Grafana(BaseController):
-
     @Endpoint()
     @ReadPermission
-    @EndpointDoc("List Grafana URL Instance",
-                 responses={200: URL_SCHEMA})
+    @EndpointDoc("List Grafana URL Instance", responses={200: URL_SCHEMA})
     def url(self):
-        response = {'instance': Settings.GRAFANA_API_URL}
+        grafana_url = mgr.get_module_option('GRAFANA_API_URL')
+        grafana_frontend_url = mgr.get_module_option('GRAFANA_FRONTEND_API_URL')
```

Comment:
I'm not sure if we want to tell users that they should use TLS certificates for IPs, which we had to if we went with the "Use IP addresses directly" proposal. PQDNs would also need to be set correctly by cephadm. If cephadm is not capable of setting it correctly, the user will need an alternative option to set it manually.

> cephadm just does not control the subject of TLS certificates. Can be anything: IPs, different domain names, different host names. We just can't demand that the ceph's internal host names used by the cluster match the TLS certificate of Grafana.
>
> Not saying anything about this implementation here, but there has to be some way for users to make the dashboard embed Grafana as defined in the certificate.

Reply:
Mmm so, in this case, whoever configuring the certs (cephadm or a manual operator) should know exactly what's the 'user-facing' URL?

I don't know exactly what's the best approach here, since I feel the problem is a bit undefined, but if we all agree on that the existing Dashboard backend-Grafana interaction is unnecessary, then we can free that configuration option and use it for this purpose (or simply provide a check to ensure the front-end and back-end options are not incompatible)...

Reply:
I don't think cephadm will ever be able to always determine the correct user-facing URL, neither as it is now nor with the knowledge  of the certificate.

In my humble opinion the best approach is to simply continue merging this PR (when I fixed the tests of course) and then to discuss different open topics around monitoring that do affect how the feature in this PR might needed to be changed. Like the re-evaluation of the reverse proxy for Grafana or the high availability. Each of these topics can change this implementation and by only looking at this implementation, I won't be able to determine the best implementation.

Reply:
I'm feeling like a broken record now. cephadm can't always determine the correct thing that the user needs to point to, in order for user's browser to reach Grafana. It just can't.

* Might be that the Grafana certificate has a specific common name
* Might be that the user need to access Grafana via a proxy
* Might be that a user need to use an FQDN to access Grafana from outside of the cluster, while the cluster itself is configured without DNS. 

Therefor users need to be able to set the grafana URL to some custom name. 

I mean, we already have a document that covers exactly that:

https://docs.ceph.com/en/latest/mgr/dashboard/#configuring-dashboard

I still don't understand why we must break user's expectations here.

Reply:
@sebastian-philipp: so we are expecting that, after the cluster is deployed with a deployment tool (cephadm), users will still need to perform several manual steps involving setting IP addresses, hostnames, etc. just for the dashboard to simply work?

I don't think that's user-friendly at all, and at least it doesn't fill the feature gap from previous Ceph deployment tools (I can only speak for Ceph-Ansible, which for example had a [`dashboard_frontend_vip` setting](https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-dashboard/tasks/configure_dashboard.yml#L185) for configuring a front-end FQDN). I don't think it's correct to embed complexity within the dashboard, when the complexity can be dealt with from the 'configuration' tool.

That said, and given this is affecting a downstream product, I already gave my ok to this PR, and just asked @p-se to add a fix so these 2 options don't allow users to make mistakes by entering incorrect data (e.g.: by configuring different HTTP schemes or ports for the same Grafana endpoint).

Reply:
> @sebastian-philipp: so we are expecting that, after the cluster is deployed with a deployment tool (cephadm), users will still need to perform several manual steps involving setting IP addresses, hostnames, etc. just for the dashboard to simply work?

The problem is, cephadm can't always determine the correct thing. It's simply outside of the control of cephadm. cephadm doesn't mandate a working DNS. cephadm doesn't mandate which url a user need to enter to access Grafana. cephadm doesn't mandate that a hostname must be a FQDN. cephadm doesn't mandate that host names must be bare. it's simply that cephadm doesn't care, as long as we can properly connect to the host. How uses access Gafana is something that cephadm doesn't care about. Sure, we can provide a good default, but that is not necessarily working for all users.

>  Ceph-Ansible, which for example had a dashboard_frontend_vip setting for configuring a front-end FQDN).

Exactly! ceph-ansible exposes a *user defined* setting `dashboard_frontend_vip` to the user to allow them to change it and is simply forwarding this to `ceph dashboard set-grafana-api-url`, because ceph-ansible has the same problem as cephadm. In constrast to ceph-ansible, cephadm doesn't provide a `dashboard_frontend_vip` setting yet, and instead let users change the grafana url themselves.

Reply:
> The problem is, cephadm can't always determine the correct thing.

Mmm, I never said that cephadm _had to guess_ this value from scratch, just that I think it should set it. Where from? Perhaps it could be inferred from another deployment setting or be directly set from a `dashboard_frontend_vip`-like option... 

My main concern about this PR was that it's providing an ad-hoc fix while leaving some other issues behind. So we'd later need to create `MGR_REDIRECT_FRONTEND_URL`, `SSO_FRONTEND_URL` and `OPENAPI_DOCS_URL` with probably the exact same data.

However, the dashboard simply guesses its hostname from the `socket.getfqdn()`, which in turn looks up `/etc/hosts`, so perhaps the [approach described here](https://tracker.ceph.com/issues/46304#note-7) might perfectly work for all those cases without the need of these changes.

Reply:
> That said, and given this is affecting a downstream product, I already gave my ok to this PR, and just asked @p-se to add a fix so these 2 options don't allow users to make mistakes by entering incorrect data (e.g.: by configuring different HTTP schemes or ports for the same Grafana endpoint).

Isn't it so that if a reverse proxy is configured neither the scheme nor the port needs to be identical with the URL the dashboard backend needs to use compared with the one frontend users would need to use?

If you agree, that for the reverse proxy to work with this setting, we cannot implement the check you've requested, can we get this PR merged then?

Reply:
Yes, you're right on that. I still miss fixing the whole issue, but as agreed, let's go and merge this one when the dashboard tests are passing.